scene/2d/tile_map.cpp

```diff
@@ -40,175 +40,291 @@
+
+		// Modulate the layer.
+		Color layer_modulate = modulate;
+		int selected_layer = tile_map_node->get_selected_layer();
+		if (selected_layer >= 0 && layer_index_in_tile_map_node != selected_layer) {
+			int z_selected = tile_map_node->get_layer_z_index(selected_layer);
+			if (z_index < z_selected || (z_index == z_selected && layer_index_in_tile_map_node < selected_layer)) {
+				layer_modulate = layer_modulate.darkened(0.5);
+			} else if (z_index > z_selected || (z_index == z_selected && layer_index_in_tile_map_node > selected_layer)) {
+				layer_modulate = layer_modulate.darkened(0.5);
+				layer_modulate.a *= 0.3;
+			}
+		}
+		rs->canvas_item_set_modulate(ci, layer_modulate);
+	}
+
+	// ----------- Quadrants processing -----------
 
-		RenderingServer *rs = RenderingServer::get_singleton();
+	// List all rendering quadrants to update, creating new ones if needed.
+	SelfList<RenderingQuadrant>::List dirty_rendering_quadrant_list;
 
-		// Free the canvas items.
-		for (const RID &ci : q.canvas_items) {
-			rs->free(ci);
+	// Check if anything changed that might change the quadrant shape.
+	// If so, recreate everything.
+	if (forced_cleanup || dirty.flags[DIRTY_FLAGS_LAYER_Y_SORT_ENABLED] || dirty.flags[DIRTY_FLAGS_TILE_MAP_QUADRANT_SIZE]) {
+		// Free all quadrants
+		for (const KeyValue<Vector2i, Ref<RenderingQuadrant>> &kv : rendering_quadrant_map) {
+			for (int i = 0; i < kv.value->canvas_items.size(); i++) {
+				const RID &ci = kv.value->canvas_items[i];
+				if (ci.is_valid()) {
+					rs->free(ci);
+				}
+			}
+			kv.value->cells.clear();
 		}
-		q.canvas_items.clear();
+		rendering_quadrant_map.clear();
+		_rendering_was_cleaned_up = true;
+	}
 
-		// Free the occluders.
-		for (const KeyValue<Vector2i, RID> &kv : q.occluders) {
-			rs->free(kv.value);
+	if (!forced_cleanup) {
+		// List all quadrants to update, recreating them if needed.
+		if (dirty.flags[DIRTY_FLAGS_TILE_MAP_TILE_SET] || _rendering_was_cleaned_up) {
+			// Update all cells.
+			for (KeyValue<Vector2i, CellData> &kv : tile_map) {
+				CellData &cell_data = kv.value;
+				_rendering_quadrants_update_cell(cell_data, dirty_rendering_quadrant_list);
+			}
+		} else {
+			// Update dirty cells.
+			for (SelfList<CellData> *cell_data_list_element = dirty.cell_list.first(); cell_data_list_element; cell_data_list_element = cell_data_list_element->next()) {
+				CellData &cell_data = *cell_data_list_element->self();
+				_rendering_quadrants_update_cell(cell_data, dirty_rendering_quadrant_list);
+			}
 		}
-		q.occluders.clear();
 
-		// Those allow to group cell per material or z-index.
-		Ref<Material> prev_material;
-		int prev_z_index = 0;
-		RID prev_ci;
+		// Update all dirty quadrants.
+		for (SelfList<RenderingQuadrant> *quadrant_list_element = dirty_rendering_quadrant_list.first(); quadrant_list_element;) {
+			SelfList<RenderingQuadrant> *next_quadrant_list_element = quadrant_list_element->next(); // "Hack" to clear the list while iterating.
```

Comment:
Would it be significantly slower to collect items to delete and then do the clearing? This feels very error prone.

Reply:
Well my question is if it's used for a good reason or because it was just simpler to write code this way 🙃 There is no natural increment in the loop's definition, so a tiny mistake can make it go on forever. Same applies to any other place that uses this trick.

Reply:
The problem is that the whole point of SelfList is to limit allocations to a minimum. If I allocate a structure to list the objects to delete, there's little gain into using a SelfList at all. 

I believe (didn't test) the performance gain likely depends on how many quadrants you have, but I guess it could get quite big with Y-sorting enabled, as the quadrant size is reduced to 1 tile = 1 quadrant.

In any case, I called that a "hack" as, like, it's a bit of a weird implementation. But TBH, I think it's perfectly fine. There's no reason the pointer should point to garbage, even if the SelfList implementation changed, as we are basically pointing an object that is allocated by the TileMap. I could remove the comment if that's a bit bothering.